### PR TITLE
[Timeseries] TS-12-TFX-Upgrade-0.23.0

### PR DESCRIPTION
**Pull request by rezarokni**
Upgrade to 0.23.0 

Remove need for config to be passed to process_fn via labels.
Ability to use tf.transform post inference.
Add sin/cos coord for timestamps.

### CODE REVIEW

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py [hunk 1]
```diff
@@ -54,6 +52,7 @@ def create_pipeline(
         ai_platform_training_args: Optional[Dict[Text, Text]] = None,
         ai_platform_serving_args: Optional[Dict[Text, Any]] = None,
         trainer_custom_config: Optional[Dict[Text, Any]] = None,
+        transformer_custom_config: Optional[Dict[Text, Any]] = None,
 ) -> pipeline.Pipeline:
     components = []
```
Comment (1025KB) [inline timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py:58]: ImportExampleGen is used below, so streaming beam data is already materialized in data_path, right?

no streaming from start of TFX just like [this](https://github.com/tensorflow/tfx/pull/2377)?

Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py:58]: Yes the data is 'continuously' materialised to the Data location. However there will be a separate PR raised for the sample to make use of a custom example gen next week. The custom example gen will be used to ensure train and test data have temporal separation.

File: timeseries-streaming/timeseries-python-applications/MLPipeline/setup.py [hunk 2]
```diff
@@ -16,16 +16,11 @@
...
-    # We need to fine-tune them and use a range of acceptable deps if possible,
-    # to be less restrictive.
+    name='timeseries-sample',
+    version='0.3.1-sample',
     install_requires=[
-      'tfx==0.22.0',
-      'kfp==0.5.1',
-      'tensorflow==2.2.0',
-      'apache-beam==2.23.0'
+      'tfx==0.23.0',
```
Comment (1025KB) [inline timeseries-streaming/timeseries-python-applications/MLPipeline/setup.py:22]: 0.24 is released

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_preprocessing.py [hunk 3]
```diff
@@ -12,70 +12,49 @@
...
     Args:
       inputs: map from feature keys to raw not-yet-transformed features.
+      custom_config:
+        timesteps: The number of timesteps in the look back window
+        features: Which of the features from the TF.Example to use in the model.
 
     Returns:
       Map from string feature key to transformed feature operations.
     """
+    timesteps = custom_config['timesteps']
```
Comment (1025KB) [inline timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_preprocessing.py:36]: this custom config is user provided? I saw the code is different [here](https://github.com/GoogleCloudPlatform/dataflow-sample-applications/blob/master/timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_preprocessing.py#L40)

which one is head?

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py [hunk 4]
```diff
@@ -29,18 +30,19 @@ def _input_fn(
...
-            # TODO Document. Shuffle must be set to False.
-            shuffle=False,
-            features=transformed_feature_spec,
-            reader=_gzip_reader_fn)
+        file_pattern=file_pattern,
+        batch_size=batch_size,
+        # Shuffle must be False, as Zip operation is not transitive
+        shuffle=False,
+        features=transformed_feature_spec,
+        reader=_gzip_reader_fn)
```
Comment (1025KB) [inline timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py:41]: you can use [tfxio](https://github.com/tensorflow/tfx/blob/master/tfx/examples/iris/iris_utils_native_keras.py#L89) to get dataset now

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py [hunk 5]
```diff
@@ -82,39 +88,38 @@ def run_fn(fn_args: TrainerFnArgs):
...
-            'serving_default': _get_serve_tf_examples_fn(
-                    model, tf_transform_output).get_concrete_function(
-                            tf.TensorSpec(
-                                    shape=(None),
-                                    dtype=tf.string,
-                                    name='examples')),
+        'serving_default': _get_serve_tf_examples_fn(
+            model, tf_transform_output).get_concrete_function(
+            tf.TensorSpec(
+                shape=None,
```
Comment (1025KB) [inline timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py:119]: in other [examples](https://github.com/tensorflow/tfx/blob/master/tfx/examples/iris/iris_utils_native_keras.py#L256) it's all set to [None], Is None correct?

File: timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py [hunk 6]
```diff
@@ -43,8 +43,9 @@
     trainer_custom_config={'epochs': 30,
```
Comment (1025KB) [inline timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py:43]: why epochs is needed? above train steps is for iteration count

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_preprocessing.py (hunk 3 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_preprocessing.py:36]: Yes with the upgrade to 0.23.0 we can now have the user pass this information into preprocessing_fn without the need to do the Key based workaround.

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py (hunk 4 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py:41]: Which is the preferred option? Any pros / cons?

File: timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py (hunk 6 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py:43]: That would set the train steps per epoch. Also is there a derived way to set this with the data from StatsGen stage? So given batch size of x derive number of training steps given epchs of y... Would be nice to have that more automated.

Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py:43]: I see the pattern used here, does the comp based on the data size: https://github.com/tensorflow/tfx/blob/master/tfx/examples/iris/iris_utils_native_keras.py
I can change the custom values to be one of data size and batch size, and then as per the iris example have the value of epoch become s=int(fn_args.train_steps / steps_per_epoch) where steps_per_epoch is  steps_per_epoch = _TRAIN_DATA_SIZE / _TRAIN_BATCH_SIZE , but pass in the _TRAIN values as custom values.

File: timeseries-streaming/timeseries-python-applications/MLPipeline/setup.py (hunk 2 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/setup.py]: Done.

File: timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py (hunk 6 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py]: Update in latest commit , please check. 1025KB@

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py (hunk 4 above)
Comment (1025KB) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py]: tfxio is perferred, we changed all major examples to use tfxio now

File: timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py (hunk 6 above)
Comment (1025KB) [reply timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py]: epochs has some [issue](https://stackoverflow.com/questions/45989971/distributed-training-with-tf-estimator-resulting-in-more-training-steps) in distributed training, that's why we all use steps in our examples

Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipelineExamples/sin_wave_example/training/timeseries_local_simple_data.py]: Right now I still populate epoch but its derived from batch / total_sample_size. Should I not set epch at all then?

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py (hunk 1 above)
Comment (ConverJens) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py:56]: @rezarokni Has this custom example gen been implemented?

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py (hunk 4 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py]: @1025KB will switch over to tfxio in a separate commit, after this branch is merged, which we will target for this week.

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py (hunk 5 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/encoder_decoder/encoder_decoder_run_fn.py]: Changed, although looks like both None and [None] are ok.

File: timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py (hunk 1 above)
Comment (rezarokni) [reply timeseries-streaming/timeseries-python-applications/MLPipeline/timeseries/pipeline_templates/timeseries_pipeline.py:56]: @ConverJens Not yet. Now that this was merged can spend more time on it.
